ompi/mca/pml/ucx/pml_ucx_datatype.c

```diff
@@ -80,13 +81,32 @@ static ucs_status_t pml_ucx_generic_datatype_unpack(void *state, size_t offset,
 
     uint32_t iov_count;
     struct iovec iov;
+    opal_convertor_t conv;
 
     iov_count    = 1;
     iov.iov_base = (void*)src;
     iov.iov_len  = length;
 
-    opal_convertor_set_position(&convertor->opal_conv, &offset);
-    opal_convertor_unpack(&convertor->opal_conv, &iov, &iov_count, &length);
+    /* in case if unordered message arrived - create separate convertor to
+     * unpack data. */
+    if (offset != convertor->offset) {
+        OBJ_CONSTRUCT(&conv, opal_convertor_t);
+        opal_convertor_copy_and_prepare_for_recv(ompi_proc_local_proc->super.proc_convertor,
+                                                 &convertor->datatype->super,
+                                                 convertor->opal_conv.count,
+                                                 convertor->opal_conv.pBaseBuf, 0,
+                                                 &conv);
+        opal_convertor_set_position(&conv, &offset);
+        opal_convertor_unpack(&conv, &iov, &iov_count, &length);
+        opal_convertor_cleanup(&conv);
+        OBJ_DESTRUCT(&conv);
+        /* permanently switch to un-ordered mode */
+        convertor->offset = 0;
+    } else {
+        opal_convertor_set_position(&convertor->opal_conv, &offset);
```

Comment:
If you are here the convertor is already set at the right position, so the call to opal_convertor_set_position is unnecessary.

Reply:
@bosilca fixed